packages/bsky/src/api/blob-resolver.ts

```diff
@@ -1,160 +1,307 @@
+
+  // Start both promises in parallel, but don't use Promise.all as we don't care
+  // about the identity result if the blob was takendown, and we want to return
+  // (or throw) ASAP.
+  const takendownPromise = dataplane.getBlobTakedown({ did, cid: cidStr })
+  const identityPromise = dataplane.getIdentityByDid({ did }, { signal })
+
+  // Avoid "uncaught promise rejection" error (we will await later, if we care)
+  identityPromise.catch(() => {})
+
+  const { takenDown } = await takendownPromise
+  // If the blob is taken down, don't wait for the identity promise
   if (takenDown) {
+    abortController.abort() // We don't care about the identity response
     throw createError(404, 'Blob not found')
   }
 
-  const blobResult = await retryHttp(() =>
-    getBlob(ctx, { pds, did, cid: cidStr }),
-  )
-  const imageStream: Readable = blobResult.data
-  const verifyCid = new VerifyCidTransform(cid)
-
-  forwardStreamErrors(imageStream, verifyCid)
-  return {
-    pds,
-    contentType:
-      blobResult.headers['content-type'] || 'application/octet-stream',
-    stream: imageStream.pipe(verifyCid),
-  }
-}
+  return identityPromise.then(
+    (identity) => {
+      const services = unpackIdentityServices(identity.services)
+      const pds = getServiceEndpoint(services, {
+        id: 'atproto_pds',
+        type: 'AtprotoPersonalDataServer',
+      })
 
-async function getBlob(
-  ctx: AppContext,
-  opts: { pds: string; did: string; cid: string },
-) {
-  const { pds, did, cid } = opts
-  return axios.get(`${pds}/xrpc/com.atproto.sync.getBlob`, {
-    params: { did, cid },
-    decompress: true,
-    responseType: 'stream',
-    timeout: 5000, // 5sec of inactivity on the connection
-    headers: getRateLimitBypassHeaders(ctx, pds),
-  })
+      if (!pds) {
+        throw createError(404, 'Origin not found')
+      }
+
+      return pds
+    },
+    (err) => {
+      if (isDataplaneError(err, Code.NotFound)) {
+        throw createError(404, 'Origin not found')
+      }
+      throw err
+    },
+  )
 }
 
-function getRateLimitBypassHeaders(
-  ctx: AppContext,
-  pds: string,
-): { 'x-ratelimit-bypass'?: string } {
-  const {
+export function* getBlobHeaders(
+  {
     blobRateLimitBypassKey: bypassKey,
     blobRateLimitBypassHostname: bypassHostname,
-  } = ctx.cfg
-  if (!bypassKey || !bypassHostname) {
-    return {}
-  }
-  const url = new URL(pds)
-  if (bypassHostname.startsWith('.')) {
-    if (url.hostname.endsWith(bypassHostname)) {
-      return { 'x-ratelimit-bypass': bypassKey }
-    }
-  } else {
-    if (url.hostname === bypassHostname) {
-      return { 'x-ratelimit-bypass': bypassKey }
+  }: ServerConfig,
+  url: URL,
+): Generator<[string, string], void, unknown> {
+  if (bypassKey && bypassHostname) {
+    const matchesUrl = bypassHostname.startsWith('.')
+      ? url.hostname.endsWith(bypassHostname)
+      : url.hostname === bypassHostname
+
+    if (matchesUrl) {
+      yield ['x-ratelimit-bypass', bypassKey]
     }
```

Comment:
I wonder if it would be a bit more straightforward to build an array then return it.